Documentation changes for features added as part of epic wazuh 7749 (#3768)
wazuh/wazuh-documentation: source/user-manual/capabilities/vulnerability-detection/how_it_works.rst

How it works
============

> To be able to detect vulnerabilities, now agents are able to natively collect a list of installed applications (:ref:`System inventory <syscollector>`), sending it periodically to the manager (where it is stored in local sqlite databases, one per agent). Also, the manager builds a global vulnerability database, from publicly available CVE repositories, using it later to cross-correlate this information with the agent's applications inventory data.

Reviewer: Maybe it's time to eliminate the **now** expression, as we aren´t explaining a new functionality when talking about syscollector.

Author: I agree

This database can be configured to be updated periodically, ensuring that the solution will check for the very latest CVEs.

Once the global vulnerability database (with the CVEs) is created, the detection process looks for vulnerable packages in the inventory databases (unique per agent). Alerts are generated when a CVE (Common Vulnerabilities and Exposures) affects a package that is known to be installed in one of the monitored hosts. A package is labeled as vulnerable when its version is contained within the affected range of a CVE.
The results are presented as alerts and also stored in a per-agent vulnerabilities inventory. So you can check the last scan alerts, or query every single agent's vulnerabilities inventory.

.. _vuln_det_scan_types:

Scan types
^^^^^^^^^^

It can run a scan on startup (:ref:`run_on_start <vuln_det_run_on_start>`) and every certain period of time (:ref:`interval <vuln_det_interval>`).

> In any of these cases, the packages and the operating system that were already scanned will not be re-scanned, only the new ones will do and alerts are going to be generated.

Reviewer 2: ```suggestion
In any of these cases, the packages and the operating system that were already scanned will not be re-scanned. Only new installed packages will be considered for the scans and alerts will be generated if they have vulnerabilities.
```

> If the database of vulnerabilities receives new CVEs information and :ref:`min_full_scan_interval <vuln_det_min_full_scan_interval>` expires all packages, as well as the operating system, will be re-scanned.

Reviewer 2: ```suggestion
There is only one case in which the packages and the operating system will be re-scanned, when the database of vulnerabilities receives new CVEs information and :ref:`min_full_scan_interval <vuln_det_min_full_scan_interval>` expires.
```

We have then three different types of scan:

- `Baseline`: This scan type will be triggered the first time after `Vulnerability Detector` is enabled. It performs a full scan for every single package installed as well as the operating system. As a result, the CVEs inventory will have the information of the vulnerabilities detected but no alerts will be generated.
- `Full scan`: In this scan type, every single package installed as well as the operating system are scanned. It runs only when the configured :ref:`min_full_scan_interval <vuln_det_min_full_scan_interval>` expires and the CVEs database is updated with new information. As a result, the vulnerabilities update will be alerted.

> - `Partial scan`: Only new packages are scanned and, as it is in `Full scans`, any vulnerability update will be reported.

Reviewer 2: ```suggestion
- `Partial scan`: Only new packages are scanned and, as it is in `Full scans`, any vulnerability update will be alerted.
```

> - `Partial scan`: Only new packages are scanned and, as it is in `Full scans`, any vulnerability update will be reported.

Reviewer: In my previous comment, I suggested adding the same explanation "as it is in `Full scans`", not to add this expression. I think is not clear. We should only explain that 
`Only new packages are scanned. As a result, the vulnerabilities update will be alerted.`

Author: Yes, I understood, I just thought it would be good to add that sentence as it is. Fixed now.

There are few considerations that arise from this behavior:

> - The `partial scan` can be triggered with a manager restart and generate alerts for new packages, as well as alerts for removed packages.

Reviewer: Not only partial scans can be triggered after a manager restart. 
i.e. The first restart will trigger a baseline.
And a full_scan can be triggered if the conditions match.

Author: Good point

> - The `partial scan` can be triggered with a manager restart and generate alerts for new packages, as well as alerts for removed packages.

Reviewer 2: Move this consideration to the `partial scan` description, without repeating the consideration about the alerts.

Author: According to @palaciosjeremias I think it would be better to not mention that consideration.

> - The new :ref:`min_full_scan_interval <vuln_det_min_full_scan_interval>` setting, will protect the manager performance from running too often `Full scans` in case of receiving many updates of vulnerabilities feeds.

Reviewer 2: Remove the `new` word.

Reviewer 2: After this, add the description of the four vulnerabilities statuses and the two types. Please, mention that they are specified in the agents' vulnerabilities inventory.

Author: Description added

Alerts generation

> ^^^^^^

Reviewer 2: Fill the remaining underline.

> - No alerts are generated for the results of a `Baseline` scan.

Reviewer 2: Remove this first bullet because it is already explained in the scan types. Instead, add a first bullet that explains the two types of alerts (new vulnerability added to the inventory and vulnerability removed from the inventory).

This database can be configured to be updated periodically, ensuring that the solution will check for the very latest CVEs.

Once the global vulnerability database (with the CVEs) is created, the detection process looks for vulnerable packages in the inventory databases (unique per agent). Alerts are generated when a CVE (Common Vulnerabilities and Exposures) affects a package that is known to be installed in one of the monitored hosts. A package is labeled as vulnerable when its version is contained within the affected range of a CVE.
The results are presented as alerts and also stored in a database. So you can check the last scan alerts, or query every single agent's vulnerable software DB.

.. _vuln_det_scan_types:

Scan types
^^^^^^^^^^

> The `Vulnerability Detector` module performs a scan when it is enabled generating a vulnerabilities inventory of packages and operating system installed in the managed agents.

Author: I agree

It also can run a scan on startup (:ref:`run_on_start <vuln_det_run_on_start>`) and every certain period of time (:ref:`interval <vuln_det_interval>`).
In any of these cases, the packages and the operating system that were already scanned will not be re-scanned, only the new ones will do and alerts are going to be generated.

> If the database of vulnerabilities receives new CVEs information and :ref:`min_full_scan_interval <vuln_det_min_full_scan_interval>` expires all packages as well as the operating system will be re-scanned.

Reviewer: Considering Diego´s suggestion in line 29, we shouldn´t mention _**In any of these cases**_. It will be clearer to not differentiate the first scan done on startup here as can be confused with the baseline functionality

> Reducing the times that the `Full scans` are triggered improves the usage of the system resources and increase the performance avoiding repeat alerts of already notified vulnerabilities during a configurable time.

Reviewer 3: Or maybe we could change to something like: "A complete scan is only run if a feed update occurred, this prevents unnecessary full scans"

> Reducing the times that the `Full scans` are triggered improves the usage of the system resources and increase the performance avoiding repeat alerts of already notified vulnerabilities during a configurable time.

Author: I considered to delete this line as below there's a explanation about the setting.

For more details we have then three different scan types that are described below.

- `Baseline`: This scan type will be triggered the first time after `Vulnerability Detector` is enabled. It performs a full scan for every single package installed as well as the operating system. As a result, the CVEs inventory will have the information of the vulnerabilities detected but no alerts will be generated.

> - `Full scan`: In this scan type, every single package installed as well as the operating system are scanned. It runs only when the configured :ref:`min_full_scan_interval <vuln_det_min_full_scan_interval>` expires and the CVEs database is updated with new information. As a result, the newly detected vulnerabilities and the ones that no longer affect the agent will be alerted.

Reviewer 3: I am making a comment here, but I am talking about the `user-manual/reference/ossec-conf/vuln-detector.html` section:

The `min_full_scan_interval` section has a note that says the following:

```
In a partial scan new packages are scanned, but only a full scan removes the CVEs related to old packages that are not present anymore.
```

But that is no longer true, a partial scan would alert of OBSOLETE vulnerabilities too.

Author: Nice, good catch. I overlooked that!

Author: Fixed!

> - `Partial scan`: Only new packages are scanned and the new vulnerabilities detected are alerted together with deleted or updated packages.

Reviewer: We should add here that removed vulnerabilities will be reported, as it is in full_scan.

Reviewer: Or, we can refer in both scans to report **"vulnerabilities update"** wrapping the addition and the deletion of vulnerabilities.
It will be explained in more detail in the **Alerts generation** section.

There are few considerations that arise from this behavior:

> - The user can not trigger a `Full scan` manually. The only option is to disable and enable `Vulnerability Detector` to trigger the `Baseline` scan. For that purpose, a service restart is needed.

Reviewer 2: Not sure whether this is totally true. We should suggest in any case the reduction of the `min_full_scan_interval`.

Reviewer 3: With the new behavior, this workaround wouldn't do anything. 
Also, a `min_full_scan_interval` reduction would only trigger a full scan if the feeds were updated.

Reviewer: This isn´t correct right now. I think it will be better to avoid mentioning the lack of functionalities and remove this paragraph

Author: I agree that would be better not to mention this.

- The `partial scan` can be triggered with a manager restart and generate alerts for new packages, as well as alerts for removed packages.
- The new :ref:`min_full_scan_interval <vuln_det_min_full_scan_interval>` setting, will protect the manager performance from running too often `Full scans` in case of receiving many updates of vulnerabilities feeds.

> Alerts

Reviewer 2: ```suggestion
Alerts generation
```

^^^^^^

- No alerts are generated for the results of a `Baseline` scan.
- Detection alerts are generated in both, `Partial scan` and `Full scan` when is detected that a vulnerability affects one of the packages scanned.

> - Removal alerts are generated in both, `Partial scan` and `Full scan`, when is detected that a vulnerability is obsolete.

Reviewer 2: Here you are mentioning the status `obsolete` of the vulnerabilities. This is not explained in the documentation and it should be because this information, as well as the type, will be returned in the API requests as well as displayed in the Wazuh UI.

Reviewer 3: These three conclusions about "Alerts" seem to be redundant.
I think that the information under "There are few considerations that arise from this behavior" already cover these facts.

Reviewer: I think we can replace the "obsolete" expression, which is an internal status of the vuln_cves table, with "deleted/updated".

Author: I agree about the "obsolete" term, I was not completely sure about deleting this block, but it's true that could be redundant in some parts. I still think it should be good idea to have an "Alerts" section.

Check :doc:`Vulnerability detector settings<../../reference/ossec-conf/vuln-detector>` for more configuration details.

> The following example diagram may be useful to understand all steps involved in the `vulnerability detector` workflow.

Reviewer: I think we can express this like
`The following example diagram may be useful to understand the steps involved in a typical `vulnerability detector` workflow.`
to reflect that this is only a simplified workflow diagram.

Author: Fixed!
